File: module-persistence/src/main/java/com/hanghae/module/persistence/repository/impl/ReservationRepositoryImpl.java
package com.hanghae.module.persistence.repository.impl;

import com.hanghae.module.domain.model.Reservation;
import com.hanghae.module.domain.repository.ReservationRepository;
import com.hanghae.module.persistence.entity.ReservationEntity;
import com.hanghae.module.persistence.mapper.ReservationMapper;
import com.hanghae.module.persistence.repository.jpa.ReservationJpaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

@Repository
@RequiredArgsConstructor
public class ReservationRepositoryImpl implements ReservationRepository {
  private final ReservationJpaRepository reservationJpaRepository;
  private final ReservationMapper reservationMapper;

  @Override
  public Reservation save(Reservation reservation) {
     ReservationEntity entity = reservationJpaRepository.save(reservationMapper.toEntity(reservation));
     return reservationMapper.toDomain(entity);
  }
}

File: infrastructure/mysql/src/main/kotlin/com/hanghe/redis/reservation/ReservationRepositoryImpl.kt
package com.hanghe.redis.reservation

import org.springframework.stereotype.Repository

@Repository
class ReservationRepositoryImpl(
    private val repository: ReservationJpaRepository,
) : ReservationRepository {

    override fun findByScreeningId(screeningId: Long): List<ReservationEntity> {
        return repository.findByScreeningId(screeningId)
    }

    override fun saveAll(newReservations: List<ReservationEntity>) {
        repository.saveAll(newReservations)
    }

    override fun findByScreeningIdAndSeatIds(screeningId: Long, seatIds: List<Long>): List<ReservationEntity> {
        return repository.findByScreeningIdAndSeatIdIn(screeningId, seatIds)
    }

    override fun findByScreeningIdAndSeatIdsWithPessimisticLock(screeningId: Long, seatIds: List<Long>): List<ReservationEntity> {
        return repository.findByScreeningIdAndSeatIdsWithPessimisticLock(screeningId, seatIds)
    }

    override fun findByScreeningIdAndSeatIdsWithOptimisticLock(
        screeningId: Long,
        seatIds: List<Long>
    ): List<ReservationEntity> {
        return repository.findByScreeningIdAndSeatIdsWithOptimisticLock(screeningId, seatIds)
    }
}
